### PR TITLE
Support filtering by interval in BAMInputFormat using a bai index file

### PR DESCRIPTION
**Pull request by tomwhite**
This is an continuation of the work in #44 by @fnothaft, but with the following changes:

* the feature is now built into BAMInputFormat/BAMRecordReader
* filters existing splits, then trims the size using chunk boundaries
* supports multiple intervals
* includes tests

### CODE REVIEW

File: src/main/java/htsjdk/samtools/PublicDiskBasedBAMFileIndex.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ NOTE!!!!
+ This is really smelly code!!!!
+
+ This class is not public in HTSJDK. We have opened a PR
+ to make the class public upstream. Once that is merged and
+ a new release of HTSJDK is cut, this file should be removed.
+ */
+public class PublicDiskBasedBAMFileIndex extends DiskBasedBAMFileIndex {
```
Comment (AndreSchumacher) [inline src/main/java/htsjdk/samtools/PublicDiskBasedBAMFileIndex.java]: Looking at https://github.com/samtools/htsjdk/commit/23eb1f02090a507ba3510f20d8b23e94a9e47516 it seems the base class is now public so maybe this is not needed anymore?

File: src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java [hunk 2]
```diff
@@ -235,6 +284,81 @@ private int addProbabilisticSplits(
...
 
+	private List<InputSplit> filterByInterval(List<InputSplit> splits, Configuration conf)
+			throws IOException {
+		List<Locatable> intervals = getIntervals(conf);
+		if (intervals == null) {
+			return splits;
+		}
+
+		// Get the chunks in the intervals we want (chunks give start and end file pointers
+		// into a BAM file) by looking in all the indexes for the BAM fils
```
Comment (AndreSchumacher) [inline src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java]: Small typo: 'BAM files'

File: src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java [hunk 3]
```diff
@@ -235,6 +284,81 @@ private int addProbabilisticSplits(
...
+			}
+			if (overlaps) {
+				filteredSplits.add(new FileVirtualSplit(virtualSplit.getPath(), newStart, newEnd,
+						virtualSplit.getLocations()));
+			}
+		}
+		return filteredSplits;
+	}
+
+	private static boolean overlaps(long start, long end, long start2, long end2) {
```
Comment (AndreSchumacher) [inline src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java]: Should we use <code>CoordMath.overlaps(..)</code> instead here as well as below?

File: src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java [hunk 4]
```diff
@@ -207,10 +213,36 @@ public static long getKey0(int refIdx, int alignmentStart0) {
 				}
 			}
 
+			if (!overlaps(r, intervals)) {
```
Comment (AndreSchumacher) [inline src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java]: We should probably keep an eye on this line whether it becomes a bottleneck when we have many intervals to filter from and then possibly switch data structures for the intervals. But let's leave it as is for now.

File: src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java [hunk 5]
```diff
@@ -207,10 +213,36 @@ public static long getKey0(int refIdx, int alignmentStart0) {
...
+
 			key.set(getKey(r));
 			record.set(r);
 			return true;
 		}
 		return false;
 	}
+
+	private static boolean overlaps(SAMRecord r, List<Locatable> intervals) {
+		if (intervals == null) {
```
Comment (AndreSchumacher) [inline src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java]: Maybe switch the condition to <pre>intervals == null || (r.getReadUnmappedFlag() && r.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START)</pre>? Then the check below could be slightly simplified but the benefit would be to completely skip interval checking for unmapped reads.

**Comment by AndreSchumacher** [comment]
@tomwhite thanks for the PR! I added a few minor comments above otherwise this looks good to me.

File: src/main/java/htsjdk/samtools/PublicDiskBasedBAMFileIndex.java (hunk 1 above)
Comment (tomwhite) [inline src/main/java/htsjdk/samtools/PublicDiskBasedBAMFileIndex.java]: Yes, a new version of htsjdk has been released now so I can remove this.

File: src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java (hunk 2 above)
Comment (tomwhite) [inline src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java]: Fixed.

File: src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java (hunk 3 above)
Comment (tomwhite) [inline src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java]: I wanted to, but it takes ints not longs, hence this method.

File: src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java (hunk 4 above)
Comment (tomwhite) [inline src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java]: OK.

File: src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java (hunk 5 above)
Comment (tomwhite) [inline src/main/java/org/seqdoop/hadoop_bam/BAMRecordReader.java]: Good point. Updated.

**Comment by tomwhite** [comment]
Addressed feedback and squashed the commits. I appreciate the review @AndreSchumacher.

**Comment by AndreSchumacher** [comment]
Thanks @tomwhite! Merging this now.

**Pull request merged**